apps/condo/domains/billing/access/BillingReceipt.js

```diff
@@ -5,31 +5,80 @@
+        return false
+    }
+    const organizationIds = serviceConsumers.map(({ organization }) => organization )
+    // Exclude all serviceConsumers for organizations without AcquiringContext in Finished status
+    const organizationsWithContract = (await find('AcquiringIntegrationContext', {
+        organization: { id_in: organizationIds, deletedAt: null },
+        integration: { deletedAt: null },
+        status: ACQUIRING_CONTEXT_FINISHED_STATUS,
+        deletedAt: null,
+    })).map(({ organization }) => organization)
+    serviceConsumers = serviceConsumers.filter(({ organization }) => organizationsWithContract.includes(organization))
+    if (!serviceConsumers.length) {
+        return false
+    }
+    const properties = await find('Property', {
+        organization: { id_in: organizationsWithContract }, deletedAt: null,
+    })
+    const propertiesForOrganizations = {}
+    properties.forEach(({ organization, addressKey }) => {
+        propertiesForOrganizations[organization] ||= []
+        propertiesForOrganizations[organization].push(addressKey)
+    })
+    return {
+        OR: serviceConsumers.map(
+            serviceConsumer => ({
+                AND: [
+                    {
+                        account: {
+                            number: serviceConsumer.accountNumber,
+                            property: { addressKey_in: propertiesForOrganizations[serviceConsumer.organization] },
+                            deletedAt: null,
+                        },
+                        context: {
+                            organization: { id: serviceConsumer.organization },
+                            deletedAt: null,
+                            status: BILLING_CONTEXT_FINISHED_STATUS,
+                        },
+                        deletedAt: null,
+                    },
+                ],
+            }),
+        ),
+    }
+}
+
 async function canReadBillingReceipts ({ authentication: { item: user } }) {
     if (!user) return throwAuthenticationError()
     if (user.deletedAt) return false
     if (user.isAdmin) return {}
 
     if (user.type === RESIDENT) {
-
-        // We don't want to make honest GQL request, as it is too expensive
-        const residents = await find('Resident', { user: { id: user.id }, deletedAt: null })
-        if (!residents || residents.length === 0) {
-            return false
-        }
-        const serviceConsumers = await find('ServiceConsumer', { resident: { id_in: residents.map(r => r.id) }, deletedAt: null })
-        if (!serviceConsumers || serviceConsumers.length === 0) {
-            return false
-        }
-
-        return {
-            OR: serviceConsumers.map(
-                s => ({ AND: [{ account: { number: s.accountNumber, deletedAt: null }, deletedAt: null, context: { id: s.billingIntegrationContext, deletedAt: null } }] } )
-            ),
-        }
+        return await buildResidentAccessToReceiptsQuery(user.id)
```

Comment:
```suggestion
        return await getResidentAccessToReceiptsQuery(user.id)
```

Reply:
You need to add a lot of tests to BillingReceipt, which are going to check your new access function

Reply:
Done